### PR TITLE
feat: add `az iot ops upgrade`

### PR DESCRIPTION
**Pull request by vilit1**
Got an M2 cluster that borks your `az iot ops list`? well worry no more! maybe

Upgrade (do not confuse with update) does the following:
- try to determine the instance version (is it m2? is it m3?)
- gets all your wonderful extensions 
- gets the sr resource id (and fails if you got a borked aio extension)
- updates the extensions in order if needed
- updates your instance if needed
- just returns if there is nothing to update
- does not make coffee for you
- it can break everything 🥇 

There is a temp hardcode btw

start
image

done with extensions, now instance
image

done done
image

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to Azure IoT Operations tooling!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT Operations CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_edge/edge/_help.py [hunk 1]
```diff
@@ -609,6 +611,30 @@ def load_iotops_help():
             az iot ops update --name myinstance -g myresourcegroup --desc "Fabrikam Widget Factory B42"
     """
 
+    helps[
+        "iot ops upgrade"
+    ] = """
+        type: command
+        short-summary: Upgrade an IoT Operations instance to the latest version.
+        long-summary: |
+                      WARNING: This command may fail and require you to delete and re-create your cluster and instance.
```
Comment (Elsie4ever) [inline azext_edge/edge/_help.py:620]: should we add this as warning while running the command?

File: azext_edge/edge/commands_edge.py [hunk 2]
```diff
@@ -134,6 +134,28 @@ def init(
...
+    instance_name: str,
+    schema_registry_resource_id: Optional[str] = None,
+    no_progress: Optional[bool] = None,
+    confirm_yes: Optional[bool] = None,
+    **kwargs
+):
+    from .providers.orchestration.upgrade import upgrade_ops_resources
+    return upgrade_ops_resources(
+        cmd=cmd,
+        # cluster_name=cluster_name,
```
Comment (Elsie4ever) [inline azext_edge/edge/commands_edge.py:149]: nit: remove if no longer needed?

File: azext_edge/edge/_help.py (hunk 1 above)
Comment (vilit1) [reply azext_edge/edge/_help.py:620]: its there already

Comment (Elsie4ever) [reply azext_edge/edge/_help.py:620]: but it's white, should warning in yellow? feel like it's a bit merged into the upgrade contexts

File: azext_edge/edge/providers/orchestration/upgrade.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        self.resource_map = self._get_resource_map()
+        # Ensure cluster exists with existing resource_map pattern.
+        self.resource_map.connected_cluster.resource
+        self.cluster_name = self.resource_map.connected_cluster.cluster_name
+
+        # get the extensions to update, populate the expected patches
+        extension_text = self._check_extensions()
+
+        if not self.extensions_to_update and not self.require_instance_upgrade:
+            logger.warning("Nothing to upgrade :)")
```
Comment (iluican) [inline azext_edge/edge/providers/orchestration/upgrade.py:97]: why warning and not info?

File: azext_edge/edge/_help.py (hunk 1 above)
Comment (vilit1) [reply azext_edge/edge/_help.py:620]: ill change the color

File: azext_edge/edge/_help.py [hunk 4]
```diff
@@ -609,6 +611,30 @@ def load_iotops_help():
...
+        examples:
+        - name: Upgrade the instance with minimal inputs.
+          text: >
+            az iot ops upgrade --name myinstance -g myresourcegroup
+        - name: Skip the conformation prompt during instance upgrade.
+          text: >
+            az iot ops upgrade --name myinstance -g myresourcegroup -y
+        - name: Upgrade the instance and specify the schema registry resource Id.
+          text: >
+            az iot ops upgrade --name myinstance -g myresourcegroup --sr-resource-id $SCHEMA_REGISTRY_RESOURCE_ID
```
Comment (c-ryan-k) [inline azext_edge/edge/_help.py:635]: If the user runs upgrade, and passes a NEW schema registry ID (not the current one) - will this just override the instance with the new version?

File: azext_edge/edge/params.py [hunk 5]
```diff
@@ -540,6 +540,15 @@ def load_iotops_arguments(self, _):
...
             )
 
+    with self.argument_context("iot ops upgrade") as context:
+        # Schema Registry
+        context.argument(
+            "schema_registry_resource_id",
+            options_list=["--sr-resource-id"],
+            help="The schema registry resource Id to use with IoT Operations. Required if the schema registry "
+            "resource Id is no longer found within IoT Operations.",
+        )
```
Comment (c-ryan-k) [inline azext_edge/edge/params.py:550]: Would it make sense to just change the previous grouping to `for cmd_space in ["iot ops init", "iot ops create", "iot ops upgrade"]:` so we only need the schema registry resource ID defined once?

Or are there other parameters used in the method header that will get picked up by the args in init and create?

File: azext_edge/edge/providers/orchestration/work.py [hunk 6]
```diff
@@ -137,12 +137,15 @@ def _format_instance_desc(self) -> str:
         return display_desc[:-1] + ""
 
     def _build_display(self):
-        pre_check_cat_desc = "Pre-Flight"
-        self._display.add_category(WorkCategoryKey.PRE_FLIGHT, pre_check_cat_desc, skipped=not self._pre_flight)
-        self._display.add_step(WorkCategoryKey.PRE_FLIGHT, WorkStepKey.REG_RP, "Ensure registered resource providers")
-        self._display.add_step(
-            WorkCategoryKey.PRE_FLIGHT, WorkStepKey.ENUMERATE_PRE_FLIGHT, "Enumerate pre-flight checks"
-        )
+        if self._pre_flight:
```
Comment (c-ryan-k) [inline azext_edge/edge/providers/orchestration/work.py:140]: If `self._pre_flight` is false, I think this is still shown in order to display that it was skipped - does this change that?

File: azext_edge/edge/_help.py (hunk 4 above)
Comment (vilit1) [reply azext_edge/edge/_help.py:635]: yup and if that breaks stuff - oh well...

File: azext_edge/edge/params.py (hunk 5 above)
Comment (vilit1) [reply azext_edge/edge/params.py:550]: the help is different so I will keep it out. Also makes it easier to remove

File: azext_edge/edge/providers/orchestration/resources/clusters.py [hunk 7]
```diff
@@ -53,3 +54,20 @@ def list(self, resource_group_name: str, cluster_name: str) -> Iterable[dict]:
             cluster_resource_name="connectedClusters",
             cluster_name=cluster_name,
         )
+
+    # will be removed
+    def update(
```
[the remark this answers is not in the document]
Comment (vilit1) [reply azext_edge/edge/providers/orchestration/resources/clusters.py:59]: yes - will rename

File: azext_edge/edge/providers/orchestration/work.py (hunk 6 above)
Comment (vilit1) [reply azext_edge/edge/providers/orchestration/work.py:140]: this was back when I was trying to stick upgrade into work

File: azext_edge/edge/providers/orchestration/upgrade.py (hunk 3 above)
Comment (vilit1) [reply azext_edge/edge/providers/orchestration/upgrade.py:97]: you know what? screw it I like print better

**Pull request merged**